### PR TITLE
[https://nvbugs/6550127][fix] Support Gemma4 multimodal cache partial hits

### PR DESCRIPTION
**Pull request by 2ez4bz**
## Dev Engineer Review

- Image and audio partial cache hits preserve item-major tensors and per-item metadata alignment.
- Unsupported layouts and modalities use the generic implementation.
- Partial video cache hits fall back to full misses and emit a one-time warning because frame-level slicing is not supported.
- Full-hit video cache reuse is also disabled. Update the related scope and TRTLLM-14981 documentation if this behavior is intentional.
- The obsolete Gemma4 NVFP4 waiver entry was removed.
- No configuration files changed.
- Verdict: needs follow-up.

## QA Engineer Review

- Updated `Gemma4MultimodalModelBase` cache tests.
- Added coverage for partial image and audio cache hits, metadata slicing, unsupported item-axis layouts, and video cache behavior.
- Video tests cover full-hit reuse and partial-hit re-encoding.
- `TestGemma4_26B_A4B::test_nvfp4` is covered by the removed waiver entry.
- No new `test-db/` or `qa/` entries are reported for the changed test functions.
- Verdict: needs follow-up.

## Description

* Why?

The generic multimodal cache path cannot slice Gemma4 image and audio layouts, causing partial encoder-cache hits to fail during input construction.

* What?

Override partial-hit input construction for Gemma4 images and audio, keeping their per-item metadata aligned. Bypass persistent caching for videos with a warning until frame-level slicing is supported, and remove the obsolete accuracy waiver.
## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Comment by BowenFu** [comment]
Two things before this lands.

**The video decision is wider than the comment says.** The comment reads "retain the pre-cache behavior", but on `main` `Gemma4MultimodalModelBase.supports_encoder_cache = True` with no `_encoder_cache_modality` override, so Gemma4 video params go through the persistent cache today and **all-hit reuse works**. What is broken is only the partial hit: the mixin's `build_multimodal_encoder_input` wants the Qwen2-VL packed layout (`video_grid_thw` + `pixel_values_videos`), Gemma4 supplies plain `pixel_values` under `"video"`, so it falls to the `else:` and raises `NotImplementedError`. Returning `None` from `_encoder_cache_modality` fixes that crash but also gives up the working all-hit path, i.e. every repeated video is re-encoded. Can you decline only the partial-hit partition and keep video cacheable for full hits? If the wider scope is deliberate, please say so in the comment and on TRTLLM-14981 — as written it reads like an edge case.

**Title is failing CI.** `Check PR Title Format` is red because the title has no `[type]` tag (and a double space): `[https://nvbugs/6550127][fix] Support Gemma4 multimodal cache partial hits`.

The rest looks right to me. The residual `MultimodalParams` dropping `multimodal_runtime` matches the mixin's own construction at `modeling_multimodal_mixin.py:705-710`, so that's the existing contract, not a new hole; the `item_count`/`dim()` guard falling through to `super()` is a good call. L0 on `889fab9` is red (63584) — worth a look at whether any of it is yours.

**Comment by xinhe-nv** [comment]
/bot run --only-qa-verify test accuracy/test_llm_api_pytorch_multimodal.py::TestGemma4_26B_A4B::test_nvfp4

**Review by StanleySun639 — APPROVED** [review]

**Review by xinhe-nv — APPROVED** [review]

**Comment by 2ez4bz** [comment]
@BowenFu addressed, please take another look.

File: tensorrt_llm/_torch/models/modeling_gemma4mm.py [hunk 1]
```diff
@@ -576,6 +582,93 @@ def multimodal_data_device_paths(self) -> List[str]:
...
+        input_tensor = modality_data.get(input_key)
+        # Only slice dim 0 after confirming it is the per-item axis declared by the cache metadata.
+        # Let the generic implementation handle any other recognized layout, or reject an
+        # inconsistent Gemma4 payload.
+        if (
+            not isinstance(input_tensor, torch.Tensor)
+            or input_tensor.dim() == 0
+            or input_tensor.shape[0] != item_count
+        ):
+            return super().build_multimodal_encoder_input(param, item_indices)
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/models/modeling_gemma4mm.py:678]: This fallback is a hard failure, not a fallback. For any Gemma4 image or audio payload, `super()` cannot match either recognized layout (image branch needs `image_sizes`, audio branch needs `input_features`), so it raises `NotImplementedError` — and `_encode_with_partial_cache` has no handler, so it propagates out of the model forward and takes down the whole iteration, not just the offending request.

So the shape guard above is really "if my assumption about dim 0 is wrong, crash the batch". Given you already have the graceful degradation pattern for video one method up, consider doing the same here: log `warning_once` and return the param unsliced is not an option, but you can bail earlier — have `partition_encoder_cache` also downgrade image/audio partitions to a full miss when `modality_data[input_key].shape[0] != len(embedding_lengths)`. That keeps the request correct under an unexpected layout (e.g. if pan-and-scan crops ever make `pixel_values.shape[0] > num_images`) instead of failing at runtime.

File: tensorrt_llm/_torch/models/modeling_gemma4mm.py [hunk 2]
```diff
@@ -576,6 +582,93 @@ def multimodal_data_device_paths(self) -> List[str]:
...
+    ) -> Optional[EncoderCachePartition]:
+        """Treat partial Gemma4 video cache hits as full misses."""
+        partition = super().partition_encoder_cache(param, encoder_cache)
+        if (
+            cls._encoder_cache_modality(param) == "video"
+            and partition is not None
+            and not partition.is_full_hit
+            and not partition.is_full_miss
+        ):
+            # Gemma4 flattens each video's frames into dim 0 before encoding, so the persistent
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/models/modeling_gemma4mm.py:601]: Concrete path for the TODO, if useful: the boundary information you need is available in `_preprocess` — the video branch has `num_soft_tokens_per_video` and `video_metadata`, and it already computes per-frame `image_seq_lens` after the `(B, F, P, C) -> (B*F, P, C)` reshape at line ~530. Persisting a `video_num_frames: List[int]` in `video_data` there would make item-level slicing a `torch.split` on the frame counts, and this override could then drop to the same shape as the image path. Might be worth noting on TRTLLM-14981 so the follow-up doesn't have to rediscover it.

File: tests/unittest/_torch/modeling/test_gemma4_multimodal.py [hunk 3]
```diff
@@ -774,6 +844,96 @@ def test_encoder_cache_reuses_image_embedding_across_requests(self):
...
+        )
+        repeated = model._get_or_encode_multimodal_embeddings(
+            [_make_keyed_video_param([shared_hash, [2] * 8])]
+        )
+
+        self.assertEqual(model.encoder_calls, 2)
+        torch.testing.assert_close(full_hit, first)
+        torch.testing.assert_close(first, torch.ones(2, model.embedding_dim))
+        torch.testing.assert_close(partial_hit, torch.full((4, model.embedding_dim), 2.0))
+        torch.testing.assert_close(repeated, partial_hit)
```
Comment (brnguyen2) [inline tests/unittest/_torch/modeling/test_gemma4_multimodal.py:968]: The test verifies the *effect* of the video downgrade but not that `partition_encoder_cache` took the intended branch — an unrelated regression that made the cache miss entirely would pass this identically. Asserting the warning fires (`self.assertLogs` / patching `logger.warning_once`) or asserting on the returned partition directly would pin the behavior. Also missing: a case where `pixel_values.shape[0] != item_count` for image/audio, which is the branch at `[modeling_gemma4mm.py:653](https://github.com/NVIDIA/TensorRT-LLM/pull/17231/files#diff-c199d40bc9ba6c34e1a2ed072172254df5e14e0d33790fae775615591b4a744fR653)` and the one with runtime-crash consequences.

File: tensorrt_llm/_torch/models/modeling_gemma4mm.py (hunk 1 above)
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_gemma4mm.py:678]: Done.

File: tensorrt_llm/_torch/models/modeling_gemma4mm.py (hunk 2 above)
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_gemma4mm.py:601]: Punting.

File: tests/unittest/_torch/modeling/test_gemma4_multimodal.py (hunk 3 above)
Comment (2ez4bz) [reply tests/unittest/_torch/modeling/test_gemma4_multimodal.py:968]: Added a test checking `logger.warning_once` was called.

**Comment by 2ez4bz** [comment]
/bot run --only-qa-verify test accuracy/test_llm_api_pytorch_multimodal.py::TestGemma4_26B_A4B::test_nvfp4
